### DISCUSSION TITLE
SPEAKコマンドでノイズがのるケースがある

### DISCUSSION THREAD

**Discussion by kaz-wan** [mmdagent-ex/MMDAgent-EX#10] [Q&A] [answered]
MMDAget-EX(Ver1.0.1)触らせていただいています。

いろいろ試させていただいているのですが、SPEAKコマンドで音声再生をすると、開始時に「パツッ」というノイズがのるケースがあるように感じました。ちなみに元の音源を聞いてみてもノイズはなさそうでした。回避策などあれば教えてください。

また別件ではありますが、TCP/IPによるソケット制御も試しているのですが、MMDAget側が同時に処理をする際など、コマンドが全て結合されて送られてきました。(MMDAget-EX側がサーバーでの利用です）
改行文字などが入っていると助かるのですが、こちらも回避策などあれば教えてください。

**Comment by LeeAkinobu** [comment, marked as answer]
利用していただきありがとうございます．

SPEAKコマンドの開始時ノイズですが，与えられたファイルはいちど内部でフォーマット変換してから再生しているので，その影響があるかもしれません．いちど .wav にして与えて，同じ症状が出るか試していただいてもよいでしょうか．また，「パツッ」が出る音声ファイルをアップしていただけたら，こちらで検証できると思います．

メッセージは1つごとに改行コードを入れていたつもりですが入っていませんでした・・・指摘ありがとうございます．先ほど修正をコミットしましたので (b9d77cd1a1feefec3139f14e2b5680c930a3bef3) 最新版を pull して試していただけますでしょうか．すべてのメッセージの末尾に改行コード \n が入ります．

**Comment by kaz-wan** [comment]
早速のご回答ありがとうございます。
改行コードに関しては改めて確認させていただきます。

ノイズが乗る件ですが、特定音源で必ず発生するというよりかは、順番の組み合わせで発生する感じがしており、ノイズ発生前に再生したものの条件によって先頭にノイズが乗る動きをしているように感じます。
テスト音源を添付しますが、必ず発生するわけではなく連続で再生したり交互にしたり、等、再生をランダムで行っていると発生しました。

[検証音源.zip](https://github.com/mmdagent-ex/MMDAgent-EX/files/14275840/default.zip)
※添付音源は22.05kHzモノラルのwav音源です

**Comment by LeeAkinobu** [comment]
テスト音源ありがとうございます．こちらでチェックして再現し，原因を特定しました．SPEAK_STARTでの音声発話終了・再開時にわずかに残った音声データを捨てていなかったことに起因するノイズでした．先ほどコードの修正を main にコミットしましたので，最新版で試していただけますでしょうか．

＃ついでに音声中の無音部でリップシンクが不要にパタつく不具合も見つけたので改善しました．

**Comment by kaz-wan** [comment]
対応ありがとうございます。
v1.0.2で確認したところ改善されておりました。
ただ、発生頻度は非常に低いのですが、末尾に同じようなノイズがのるケースがまだあるようです。
再現頻度が下がったため、確定で発生させる音源がまだ見つかっていません。
こちらもご確認いただけますでしょうか。

続きで別件になってしまうのですが、以下についても教えていただくことはできますでしょうか？

１．MODEL_CHANGEを行うと、もともとの座標から移動をしてしまうのですが、仕様でしょうか？
２．MMDAgent-EX側をサーバーとして利用している際、MMDAgent側が落ちたことを即時に知る方法などはありますでしょうか？(定期ポーリングによる死活監視の方法があるなど)

上記はタイトルと内容が離れてきているため、必要でしたら別の議題として立ち上げしたほうがいいでしょうか？
よろしくお願いいたします。

**Reply by LeeAkinobu** [reply]
ありがとうございます。SPEAK末端にノイズが乗ることがある件、issue として挙げていただけるとありがたいです（日本語でよいです）

あとの質問についてコメントします（このままでよいです）

MODEL_CHANGE で座標が変わるのはバグかもしれません。そのモデルを MODEL_ADD するさいに原点以外の座標を表示位置に指定した場合、MOVE_START 等で位置を変えた場合、再生中の MOTION の中に reposition が ON のものがある場合などの条件でそういった状況が生じる可能性はあります。MODEL_ADDとMOTION_ADD、その後のMODEL_CHANGEのコマンドのパラメータを教えていただければ調査できるかもしれません。

サーバのMMDAgent-EXが落ちたことを即時に知るのは、TCP/IP でソケット接続してソケットエラーを監視、かなと思います。あるいは、MMDAgent-EX が起動している間は[動作テスト用に http の 50000 番ポートへのアクセスを受け付けている](https://mmdagent-ex.dev/ja/docs/message-test/)ので、そこへ http を定期的に飛ばしてエラーが出たら死んだ、という判断はできるかもしれません。

**Reply by kaz-wan** [reply]
> ありがとうございます。SPEAK末端にノイズが乗ることがある件、issue として挙げていただけるとありがたいです（日本語でよいです）

こちらはもう少し再現方法を確認した後記載いたします。

>MODEL_ADDとMOTION_ADD、その後のMODEL_CHANGEのコマンドのパラメータを教えていただければ調査できるかもしれません。

こちらはシナリオ上、以下の流れになっています。

MODEL_ADD|idle_panel|XXXX.pmd|4.0,11.0,-6.0|0.0,0.0,0.0|OFF
↓
MODEL_CHANGE|idle_panel|YYYY.pmd

>サーバのMMDAgent-EXが落ちたことを即時に知るのは～

ありがとうございます。教えていただいた内容で検討したいと思います。

**Reply by LeeAkinobu** [reply]
MODEL_CHANGE で場所がリセットされる件，バグであることを確認できたのでさきほどのコミット (ae186f7972a04fbc4753b52090d4d9ab57449825) で修正しました．ご報告ありがとうございました！

**Reply by kaz-wan** [reply]
ご対応ありがとうございます。私が確認したかったことは以上となります。

最後となりましたが、MMDAgent-EXはMMDAgentに欲しいと思っていた機能がいろいろ実装されており、うれしく思っております（SPEAKコマンドや外部通信制御機能の標準装備、fstの書き方等）
引き続き今後のアップデートを楽しみにしております。

**Reply by LeeAkinobu** [reply]
ありがとうございます。これからもよろしくお願いします。
